### PR TITLE
[SPARK-30183][SQL] Disallow to specify reserved properties in CREATE/ALTER NAMESPACE syntax

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
Currently, COMMENT and LOCATION are reserved properties for Datasource v2 namespaces. They can be set via specific clauses and via properties. And the ones specified in clauses take precede of properties. Since they are reserved, which means they are not able to visit directly. They should be used in COMMENT/LOCATION clauses ONLY.

### Why are the changes needed?
make reserved properties be reserved.

### Does this PR introduce any user-facing change?
yes, 'location', 'comment' are not allowed use in db properties

### How was this patch tested?
UNIT tests.

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan, I think this a separate issue from `SET OWNER`. As to `COMMENT`, methinks if have defined it as a reserved property, it should be treat as same as `LOCATION`. Thanks for reviewing this in advance.

**Comment by cloud-fan** [comment]
The problem is, we don't have a specific clause to alter COMMENT, but I do agree that it's weird to write both COMMENT and the "comment" property in CREATE TABLE.

We should either treat comment specially, and allow to set it via properties, or create a special clause for altering comment. What does presto and RDBMS do?

**Comment by yaooqinn** [comment]
presto has no comment as database attribute https://prestodb.io/docs/current/sql/create-schema.html
postgresql has `comment on` syntax https://www.postgresql.org/docs/9.1/sql-comment.html

Mysql allows comments only for tables and columns. `ALTER TABLE t1 COMMENT = 'New table comment';` https://dev.mysql.com/doc/refman/8.0/en/alter-database.html

**Comment by cloud-fan** [comment]
presto also has the COMMENT ON command: https://prestosql.io/docs/current/sql/comment.html

Can we add it first? Then it's safe to say that these properties are reserved.

**Comment by yaooqinn** [comment]
OK, I'll add that first

**Comment by yaooqinn** [comment]
Hi @cloud-fan, do we need an umbrella for the `COMMENT ON` syntaxes? According to PostgreSQL we may have several matched cases. https://www.postgresql.org/docs/9.1/sql-comment.html

**Comment by cloud-fan** [comment]
I think we only need to support TABLE and NAMESPACE

**Comment by yaooqinn** [comment]
How about COLUMN / VIEW?

**Comment by cloud-fan** [comment]
We have SQL syntax to set comment of a column, and view is not supported by v2 yet. At least we only need to support table and namespace in the near future.

**Comment by yaooqinn** [comment]
I got it.

**Comment by yaooqinn** [comment]
Applied the new resolving framework, cc @cloud-fan, thanks for reviewing

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -2525,17 +2526,21 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     var properties = ctx.tablePropertyList.asScala.headOption
       .map(visitPropertyKeyValues)
       .getOrElse(Map.empty)
-    Option(ctx.comment).map(string).map {
-      properties += SupportsNamespaces.PROP_COMMENT -> _
+
+    if (properties.keySet.intersect(RESERVED_PROPERTIES.asScala.toSet).nonEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: shall we do the check inside `CreateNamespace` command? in case we have other APIs that can create `CreateNamespace` command in the future.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -2525,17 +2526,21 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
+      throw new ParseException(s"Disallow to specify reserved properties, including" +
+        s" ${RESERVED_PROPERTIES.asScala.mkString("(", ",", ")")}.", ctx)
+    }
+
+    Option(ctx.comment).map(string).foreach {
+      properties += PROP_COMMENT -> _
     }
-    ctx.locationSpec.asScala.headOption.map(visitLocationSpec).map {
-      properties += SupportsNamespaces.PROP_LOCATION -> _
+    ctx.locationSpec.asScala.headOption.map(visitLocationSpec).foreach {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit: why `ctx.locationSpec` is a list but `ctx.comment` isn't?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala [hunk 3]
```diff
@@ -241,10 +241,11 @@ case class ReplaceTableAsSelect(
  * The logical plan of the CREATE NAMESPACE command that works for v2 catalogs.
  */
 case class CreateNamespace(
-    catalog: SupportsNamespaces,
-    namespace: Seq[String],
+    child: LogicalPlan,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala]: can we delay this change? We may want to create the rule to convert to v1 commands when migrating to the new framework. Let's focus on disallowing reserved properties in this PR.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: `locationSpec` is different from `comment`, shall we create a `commentSpec` as `locationSpec`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 4]
```diff
@@ -2525,11 +2526,17 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     var properties = ctx.tablePropertyList.asScala.headOption
       .map(visitPropertyKeyValues)
       .getOrElse(Map.empty)
-    Option(ctx.comment).map(string).map {
-      properties += SupportsNamespaces.PROP_COMMENT -> _
+
+    if (properties.keySet.intersect(RESERVED_PROPERTIES.asScala.toSet).nonEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit:
```
properties.keySet.exists(RESERVED_PROPERTIES.contains)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: let's leave it as it is for now

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 5]
```diff
@@ -25,6 +25,7 @@ import org.apache.spark.sql.catalyst.catalog.{BucketSpec, CatalogTable, CatalogT
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.connector.catalog.{CatalogManager, CatalogPlugin, LookupCatalog, SupportsNamespaces, Table, TableCatalog, TableChange, V1Table}
+import org.apache.spark.sql.connector.catalog.CatalogV2Util.isSessionCatalog
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: unnecessary change

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 6]
```diff
@@ -2525,11 +2526,17 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     var properties = ctx.tablePropertyList.asScala.headOption
       .map(visitPropertyKeyValues)
       .getOrElse(Map.empty)
-    Option(ctx.comment).map(string).map {
-      properties += SupportsNamespaces.PROP_COMMENT -> _
+
+    if (properties.keySet.exists(RESERVED_PROPERTIES.contains)) {
+      throw new ParseException(s"Disallow to specify reserved properties, including" +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: we can make the error message better
```
properties.foreach { case (key, _)
  if (RESERVED_PROPERTIES.contains(key)) {
    fail(s"$key is a reserved properties.")
  }
}
```

**Comment by cloud-fan** [comment]
shall we add the check to ALTER NAMESPACE in this PR as well? And this is a breaking change we should add an entry in migration guide.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: after some more thought, I think this is a breaking change. Existing queries may already use the "comment" property. Can we add a legacy config and update migration guide?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: OK

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: we have made an undocumented breaking change here, the `comment` and `location` have side-effects now. Simply forbid them seems to make the influence smaller. Otherwise for fully backward compatibility, we shoud add addtional logic to let them have no side effects.
```sql
create database spark2 with dbproperites ('comment'='ct', 'location'='loc');
```
```sql
Database Name	spark2
Description
Location	file:/Users/kentyao/Downloads/spark/spark-2.4.4-bin-hadoop2.7/spark-warehouse/spark2.db
Properties	((comment,ct), (location,loc))
Time taken: 0.022 seconds, Fetched 4 row(s)
```
```sql
create database spark3 with dbproperties ('comment'='ct', 'location'='loc');
```
```sql
Database Name	spark3
Description	ct
Location	file:/Users/kentyao/Downloads/spark/spark-3.0.0-SNAPSHOT-bin-20200103/loc
Owner Name	kentyao
Owner Type	USER
Properties
Time taken: 0.019 seconds, Fetched 6 row(s)
```

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: > we can make the error message better
> 
> ```
> properties.foreach { case (key, _)
>   if (key == PROP_COMMENT) {
>     fail(s"$key is a reserved namespace property, please use the COMMENT clause to specify it.")
>   }
>   ...
> }
> ```

now we have 2 reserved props, after ownership, there shall be 2, what if the number keeps growing :)

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: ah so in Spark 2, the reserved properties are simply ignored?

BTW this error message is for end users, we should try our best to make it helpful, even if it means more code.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: the reserved properties are not simply ignored but as user-side properties, I guess

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I think it's OK that the legacy config can only ignore the reserved properties but not treat them as user-side properties. At least the query won't fail.

File: docs/sql-migration-guide.md [hunk 7]
```diff
@@ -264,6 +264,8 @@ license: |
...
 
+  - Since Spark 3.0, the namespace properties `location` and `comment` become reserved, it will fail with `ParseException` if we use them as members of `DBPROTERTIES` in `CREATE NAMESPACE` and `ALTER NAMESPACE ... SET PROPERTIES(...)`. We need their specific clauses to specify them, e.g. `CREATE NAMESPACE a.b.c COMMENT 'any comment' LOCATION 'some path'`. We can set `spark.sql.legacy.property.nonReserved` to `true` to ignore the `ParseException`, but notice that in this case, these properties will produce side effects, e.g `SET DBPROTERTIES('location'='/tmp')` might change the location of the database. In Spark version 2.4 and earlier, these properties are neither reserved nor have side effects, e.g.   `SET DBPROTERTIES('location'='/tmp')` will not change the location of the database but create a headless property just like `'a'='b'`.
+
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: hey, we should ignore them (i.e. remove them from properties), so that they won't have side effect.

File: docs/sql-migration-guide.md [hunk 8]
```diff
@@ -264,6 +264,8 @@ license: |
...
 
+  - Since Spark 3.0, the namespace properties `location` and `comment` become reserved, it will fail with `ParseException` if we use them as members of `DBPROTERTIES` in `CREATE NAMESPACE` and `ALTER NAMESPACE ... SET PROPERTIES(...)`. We need their specific clauses to specify them, e.g. `CREATE NAMESPACE a.b.c COMMENT 'any comment' LOCATION 'some path'`. We can set `spark.sql.legacy.property.nonReserved` to `true` to ignore the `ParseException`, but notice that in this case, these properties will produce side effects, e.g `SET DBPROTERTIES('location'='/tmp')` might change the location of the database. In Spark version 2.4 and earlier, these properties are neither reserved nor have side effects, e.g.   `SET DBPROTERTIES('location'='/tmp')` will not change the location of the database but create a headless property just like `'a'='b'`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: Since it's the migration guide, I think it's better to use the syntax of Spark 2.4
```
it will fail with `ParseException` if we use them in `CREATE DATABASE ... DBPROPERTIES` and `ALTER DATABASE ... SET DBPROPERTIES` ...
```

File: docs/sql-migration-guide.md (hunk 7 above)
Comment (yaooqinn) [reply docs/sql-migration-guide.md]: what I have done here, lol..

File: docs/sql-migration-guide.md [hunk 9]
```diff
@@ -264,6 +264,8 @@ license: |
...
 
+  - Since Spark 3.0, the words `location` and `comment` become reserved database properties, it will fail with `ParseException` if we use them as members of `DBPROTERTIES` in `CREATE DATABASE` and `ALTER DATABASE ... SET PROPERTIES(...)`. We need their specific clauses to specify them, e.g. `CREATE DATABASE test COMMENT 'any comment' LOCATION 'some path'`. We can set `spark.sql.legacy.property.nonReserved` to `true` to ignore the `ParseException`, but notice that in this case, these properties will be ignored too, e.g `SET DBPROTERTIES('location'='/tmp')` will affect nothing. In Spark version 2.4 and earlier, these properties are neither reserved nor have side effects, e.g. `SET DBPROTERTIES('location'='/tmp')` will not change the location of the database but only create a headless property just like `'a'='b'`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: `notice that in this case, these properties will be ignored too ...` We don't need to highlight it as it's the same with 2.4

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 10]
```diff
@@ -2498,6 +2498,26 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
   }
 
+  private def checkNamespaceProperties(
+      properties: Map[String, String],
+      ctx: ParserRuleContext): Map[String, String] = withOrigin(ctx) {
+    import SupportsNamespaces._
+    if (!conf.getConf(SQLConf.LEGACY_PROPERTY_NON_RESERVED)) {
+      properties.foreach {
+        case (PROP_LOCATION, _) =>
+          throw new ParseException(s"$PROP_LOCATION is a reserved" + s" namespace property," +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: why break down the string here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 11]
```diff
@@ -2498,6 +2498,26 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     }
   }
 
+  private def checkNamespaceProperties(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: it's not just check now, how about `cleanNamespaceProperties`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 12]
```diff
@@ -2135,6 +2135,15 @@ object SQLConf {
         "defined by `from` and `to`.")
       .booleanConf
       .createWithDefault(false)
+
+  val LEGACY_PROPERTY_NON_RESERVED =
+    buildConf("spark.sql.legacy.property.nonReserved")
+      .internal()
+      .doc("When true, all database and table properties are not reserved and available for " +
+        "create/alter syntaxes. But please be aware that the reserved properties will still be " +
+        "used by Spark internally and will ignore their user specified values.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `the reserved properties will be silently removed.`

File: docs/sql-migration-guide.md (hunk 9 above)
Comment (yaooqinn) [reply docs/sql-migration-guide.md]: not exactly the same, 2.4 will create a dbproperty, we will not do that. I'd change this to your suggestion as is in `SQLConf`

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 13]
```diff
@@ -864,6 +865,33 @@ class DataSourceV2SQLSuite
...
+          sql(s"CREATE NAMESPACE testcat.reservedTest WITH DBPROPERTIES('$key'='dummyVal')")
+        }
+        assert(exception.getMessage.contains(s"$key is a reserved namespace property"))
+      }
+    }
+    withSQLConf((SQLConf.LEGACY_PROPERTY_NON_RESERVED.key, "true")) {
+      SupportsNamespaces.RESERVED_PROPERTIES.asScala.foreach { key =>
+        withNamespace("testcat.reservedTest") {
+          withTempDir { tmpDir =>
+            val noEffectVal = tmpDir.getCanonicalPath
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: since it's no effect, can we just use a dumb string like "foo"?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 14]
```diff
@@ -961,6 +989,37 @@ class DataSourceV2SQLSuite
...
+          }
+          assert(exception.getMessage.contains(s"$key is a reserved namespace property"))
+        }
+      }
+    }
+    withSQLConf((SQLConf.LEGACY_PROPERTY_NON_RESERVED.key, "true")) {
+      SupportsNamespaces.RESERVED_PROPERTIES.asScala.foreach { key =>
+        withNamespace("testcat.reservedTest") {
+          withTempDir { tmpDir =>
+            val noEffectVal = tmpDir.getCanonicalPath
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: ditto

File: docs/sql-migration-guide.md [hunk 15]
```diff
@@ -264,6 +264,8 @@ license: |
...
 
+  - Since Spark 3.0, the words `location` and `comment` become reserved database properties, it will fail with `ParseException` if we use them as members of `DBPROTERTIES` in `CREATE DATABASE` and `ALTER DATABASE ... SET PROPERTIES(...)`. We need their specific clauses to specify them, e.g. `CREATE DATABASE test COMMENT 'any comment' LOCATION 'some path'`. We can set `spark.sql.legacy.property.nonReserved` to `true` to ignore the `ParseException`, in this case, these properties will be silently removed, e.g `SET DBPROTERTIES('location'='/tmp')` will affect nothing. In Spark version 2.4 and earlier, these properties are neither reserved nor have side effects, e.g. `SET DBPROTERTIES('location'='/tmp')` will not change the location of the database but only create a headless property just like `'a'='b'`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: nit:
```
Since Spark 3.0, `location` and `comment` become reserved database properties. Commands will fail
if we specify reserved properties in `CREATE DATABASE ... DBPROTERTIES ` and `ALTER DATABASE ... SET PROPERTIES`. ...
```

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 13 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: fixed, thanks

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
